crates/holochain/tests/signals/mod.rs

```diff
@@ -0,0 +1,135 @@
+//! Tests for local and remote signals using rendezvous config
+//!
+
+use fallible_iterator::FallibleIterator;
+use futures::StreamExt;
+use hdk::prelude::ExternIO;
+use holochain::sweettest::{SweetCell, SweetConductorBatch, SweetConductorConfig, SweetDnaFile};
+use holochain_types::signal::Signal;
+use holochain_wasm_test_utils::TestWasm;
+use holochain_zome_types::RemoteSignal;
+use serde::{Deserialize, Serialize};
+use std::time::Duration;
+
+#[derive(Serialize, Deserialize, Debug)]
+struct SignalMessage {
+    value: String,
+}
+
+fn to_signal_message(signal: Signal) -> SignalMessage {
+    match signal {
+        Signal::App { signal, .. } => signal.into_inner().decode().unwrap(),
+        _ => {
+            panic!("Only expected app signals");
+        }
+    }
+}
+
+#[tokio::test(flavor = "multi_thread")]
+#[cfg(feature = "slow_tests")]
+async fn remote_signals_batch() -> anyhow::Result<()> {
+    holochain_trace::test_run().ok();
+
+    let mut conductors =
+        SweetConductorBatch::from_config_rendezvous(2, SweetConductorConfig::rendezvous()).await;
+
+    let dna_file = SweetDnaFile::unique_from_test_wasms(vec![TestWasm::EmitSignal])
+        .await
+        .0;
+
+    let app_batch = conductors.setup_app("app", &[dna_file]).await.unwrap();
+
+    let ((alice,), (bob,)): ((SweetCell,), (SweetCell,)) = app_batch.into_tuples();
+
+    // Make sure the conductors are talking to each other before sending signals.
+    conductors
+        .require_initial_gossip_activity_for_cell(&bob, Duration::from_secs(90))
```

Comment:
Is this or should this be required prior to receiving signals?

Reply:
Hmm, it doesn't seem to be a problem for me to remove it locally but this is here to make sure that agent discovery has taken place. It's the equivalent of `exchange_peer_info`

Reply:
`exchange_peer_info` writes all agent infos into peer dbs and `require_initial_gossip_activity_for_cell` awaits the first gossip round to have completed, so I thought using the rendezvous config is equivalent to `exchange_peer_info`. Hence my question. I don't object to leaving this in at all, I'm asking because I noticed in Tryorama that ever since using local bootstrap services I didn't need to exchange peer infos any longer.

Reply:
I worked on a few flaky tests around this so my understanding of this may not be complete but the way I understand this is:

Using the rendezvous config will set up local services including a bootstrap server which means peers have a way to discover each other. This means you don't need to, and actually *shouldn't*, `exchange_peer_info` because messing with the peer data store risks conflicting with the information being bootstrapped.

Now for the bootstrapping, that is initially *super fast*. It runs multiple times in the first few seconds of the test and peers should find each other. However, when running on slower machines and running multiple conductors in one test it can take some time for one of them to start up. The back-off on the bootstrap loop is pretty aggressive (2x multiplier) so quite quickly peer discovery can become slow for conductors that were slow to start.

What I'm getting at here is that the bootstrap server gives a way for all peers to get in touch and given a bit of time they will all find each other. What isn't guaranteed is that by the time you ask for a remote signal to be sent to a given agent, that the sending conductor has discovered the remote agent to send to.

Nothing in the sweettest setup of rendezvous actually makes sure everything is bootstrapped and talking. That's why I added this `require_initial_gossip_activity_for_cell` because waiting for genesis records to be gossiped is a simple way to observe that all conductors are talking to each other.

So I think leaving this in will prevent strange test failures where remote signals don't get sent. In the real world you wouldn't have an agent key to send a remote signal to until you've discovered that peer which makes things a bit simpler, but in tests we have information sooner than we 'should'.